File: Levenshtein_Algorithm/Brute Force/BF.py
class LAResult:
    def __init__(self, cost=0, operations=None):
        self.cost = cost
        self.operations = operations if operations is not None else []


class OperationCosts:
    def __init__(self, insert_cost=1, delete_cost=1, replace_cost=1):
        self.insert_cost = insert_cost
        self.delete_cost = delete_cost
        self.replace_cost = replace_cost


def minimum_cost(a, b, c):
    return min(a, b, c)


# Recursion Method - Return only the cost
# Time Complexity: O(3^N)
# Space Complexity: O(N) {stack}
def levenshtein_cost_only(str1, str2, i, j, costs):
    
    if i == len(str1):
        return (len(str2) - j) * costs.insert_cost
    if j == len(str2):
        return (len(str1) - i) * costs.delete_cost
    
    if str1[i] == str2[j]:
        return levenshtein_cost_only(str1, str2, i + 1, j + 1, costs)
    
    insert_cost = costs.insert_cost + levenshtein_cost_only(str1, str2, i, j + 1, costs)
    delete_cost = costs.delete_cost + levenshtein_cost_only(str1, str2, i + 1, j, costs)
    replace_cost = costs.replace_cost + levenshtein_cost_only(str1, str2, i + 1, j + 1, costs)
    
    return minimum_cost(insert_cost, delete_cost, replace_cost)


# Recursion Method - Return cost + operations
# Time Complexity: O(3^N)
# Space Complexity: O(N^2) {stack + operations}
def levenshtein_cost_and_operation(str1, str2, i, j, costs):
    
    if i == len(str1):
        ops = []
        total_cost = 0
        for k in range(j, len(str2)):
            ops.append(f"Insert '{str2[k]}' (cost: {costs.insert_cost})")
            total_cost += costs.insert_cost
        return LAResult(total_cost, ops)
    
    if j == len(str2):
        ops = []
        total_cost = 0
        for k in range(i, len(str1)):
            ops.append(f"Delete '{str1[k]}' (cost: {costs.delete_cost})")
            total_cost += costs.delete_cost
        return LAResult(total_cost, ops)
    
    if str1[i] == str2[j]:
        result = levenshtein_cost_and_operation(str1, str2, i + 1, j + 1, costs)
        result.operations.insert(0, f"Match '{str1[i]}' (cost: 0)")
        return result
    
    insert_op = levenshtein_cost_and_operation(str1, str2, i, j + 1, costs)
    insert_op.cost += costs.insert_cost
    insert_op.operations.insert(0, f"Insert '{str2[j]}' (cost: {costs.insert_cost})")
    
    delete_op = levenshtein_cost_and_operation(str1, str2, i + 1, j, costs)
    delete_op.cost += costs.delete_cost
    delete_op.operations.insert(0, f"Delete '{str1[i]}' (cost: {costs.delete_cost})")
    
    replace_op = levenshtein_cost_and_operation(str1, str2, i + 1, j + 1, costs)
    replace_op.cost += costs.replace_cost
    replace_op.operations.insert(0, f"Replace '{str1[i]}' with '{str2[j]}' (cost: {costs.replace_cost})")
    
    if insert_op.cost <= delete_op.cost and insert_op.cost <= replace_op.cost:
        return insert_op
    elif delete_op.cost <= insert_op.cost and delete_op.cost <= replace_op.cost:
        return delete_op
    else:
        return replace_op


def get_costs():
    print("\n--- Operation Cost Configuration ---")
    same_cost_choice = int(input("Do all operations have the same cost? (yes=1 / no=0): "))
    
    if same_cost_choice == 1:
        uniform_cost = int(input("Enter the cost for all operations (insert, delete, replace): "))
        print(f"All operations will have cost: {uniform_cost}\n")
        return OperationCosts(uniform_cost, uniform_cost, uniform_cost)
    else:
        insert_cost = int(input("Enter the cost for insert operation: "))
        delete_cost = int(input("Enter the cost for delete operation: "))
        replace_cost = int(input("Enter the cost for replace operation: "))
        
        return OperationCosts(insert_cost, delete_cost, replace_cost)


def edit_distance(s1, s2, costs):
    print("Do you want detailed operations? (yes=1 / no=0)")
    flag = int(input("Enter your choice (1/0): "))
    
    if flag == 1:
        result = levenshtein_cost_and_operation(s1, s2, 0, 0, costs)
        print(f"\nOperations to convert '{s1}' to '{s2}':")
        for op in result.operations:
            print(f" - {op}")
        print()
        return result.cost
    
    return levenshtein_cost_only(s1, s2, 0, 0, costs)


def main():
    print("This is a Python program to calculate the Levenshtein distance with custom operation costs.\n\n")
    
    num_inputs = int(input("Enter the number of string pairs you want to process: "))
    print()
    
    for i in range(1, num_inputs + 1):
        print(f"\n\n\n ***** ***** ***** ***** Pair {i} ***** ***** ***** ***** \n")
        
        # Get operation costs configuration
        costs = get_costs()
        
        str1 = input("Enter the first string (The original string): ")
        str2 = input("Enter the second string (String to convert to): ")
        
        print(f"\nResults for {{ Pair {i} }}")
        res = edit_distance(str1, str2, costs)
        print(f"Total Edit Distance for Pair {i}: {res}")


if __name__ == "__main__":
    main()

File: Levenshtein_Algorithm/Brute Force/BF.java
import java.util.*;

class LAResult {
    int cost;
    List<String> operations;
    
    public LAResult() {
        this.cost = 0;
        this.operations = new ArrayList<>();
    }
    
    public LAResult(int cost, List<String> operations) {
        this.cost = cost;
        this.operations = operations;
    }
}

class OperationCosts {
    int insertCost;
    int deleteCost;
    int replaceCost;
    
    public OperationCosts() {
        this.insertCost = 1;
        this.deleteCost = 1;
        this.replaceCost = 1;
    }
    
    public OperationCosts(int ins, int del, int rep) {
        this.insertCost = ins;
        this.deleteCost = del;
        this.replaceCost = rep;
    }
}

public class Main {
    
    private static int minimumCost(int a, int b, int c) {
        return Math.min(a, Math.min(b, c));
    }
    
    // Recursion Method - Return only the cost
    // Time Complexity: O(3^N)
    // Space Complexity: O(N) {stack}
    private static int levenshteinCostOnly(String str1, String str2, int i, int j, OperationCosts costs) {
        
        if (i == str1.length()) {
            return (str2.length() - j) * costs.insertCost;
        }
        if (j == str2.length()) {
            return (str1.length() - i) * costs.deleteCost;
        }
        
        if (str1.charAt(i) == str2.charAt(j)) {
            return levenshteinCostOnly(str1, str2, i + 1, j + 1, costs);
        }
        
        int insertCost = costs.insertCost + levenshteinCostOnly(str1, str2, i, j + 1, costs);
        int deleteCost = costs.deleteCost + levenshteinCostOnly(str1, str2, i + 1, j, costs);
        int replaceCost = costs.replaceCost + levenshteinCostOnly(str1, str2, i + 1, j + 1, costs);
        
        return minimumCost(insertCost, deleteCost, replaceCost);
    }
    
    // Recursion Method - Return cost + operations
    // Time Complexity: O(3^N)
    // Space Complexity: O(N^2) {stack + operations}
    private static LAResult levenshteinCostAndOperation(String str1, String str2, int i, int j, OperationCosts costs) {
        
        if (i == str1.length()) {
            List<String> ops = new ArrayList<>();
            int totalCost = 0;
            for (int k = j; k < str2.length(); k++) {
                ops.add("Insert '" + str2.charAt(k) + "' (cost: " + costs.insertCost + ")");
                totalCost += costs.insertCost;
            }
            return new LAResult(totalCost, ops);
        }
        
        if (j == str2.length()) {
            List<String> ops = new ArrayList<>();
            int totalCost = 0;
            for (int k = i; k < str1.length(); k++) {
                ops.add("Delete '" + str1.charAt(k) + "' (cost: " + costs.deleteCost + ")");
                totalCost += costs.deleteCost;
            }
            return new LAResult(totalCost, ops);
        }
        
        if (str1.charAt(i) == str2.charAt(j)) {
            LAResult res = levenshteinCostAndOperation(str1, str2, i + 1, j + 1, costs);
            res.operations.add(0, "Match '" + str1.charAt(i) + "' (cost: 0)");
            return res;
        }
        
        LAResult insertOp = levenshteinCostAndOperation(str1, str2, i, j + 1, costs);
        insertOp.cost += costs.insertCost;
        insertOp.operations.add(0, "Insert '" + str2.charAt(j) + "' (cost: " + costs.insertCost + ")");
        
        LAResult deleteOp = levenshteinCostAndOperation(str1, str2, i + 1, j, costs);
        deleteOp.cost += costs.deleteCost;
        deleteOp.operations.add(0, "Delete '" + str1.charAt(i) + "' (cost: " + costs.deleteCost + ")");
        
        LAResult replaceOp = levenshteinCostAndOperation(str1, str2, i + 1, j + 1, costs);
        replaceOp.cost += costs.replaceCost;
        replaceOp.operations.add(0, "Replace '" + str1.charAt(i) + "' with '" + str2.charAt(j) + "' (cost: " + costs.replaceCost + ")");
        
        if (insertOp.cost <= deleteOp.cost && insertOp.cost <= replaceOp.cost) {
            return insertOp;
        } else if (deleteOp.cost <= insertOp.cost && deleteOp.cost <= replaceOp.cost) {
            return deleteOp;
        } else {
            return replaceOp;
        }
    }
    
    private static OperationCosts getCosts() {
        Scanner scanner = new Scanner(System.in);
        
        System.out.println("\n--- Operation Cost Configuration ---");
        System.out.print("Do all operations have the same cost? (yes=1 / no=0): ");
        int sameCostChoice = scanner.nextInt();
        
        if (sameCostChoice == 1) {
            System.out.print("Enter the cost for all operations (insert, delete, replace): ");
            int uniformCost = scanner.nextInt();
            System.out.println("All operations will have cost: " + uniformCost + "\n");
            return new OperationCosts(uniformCost, uniformCost, uniformCost);
        } else {
            System.out.print("Enter the cost for insert operation: ");
            int insertCost = scanner.nextInt();
            System.out.print("Enter the cost for delete operation: ");
            int deleteCost = scanner.nextInt();
            System.out.print("Enter the cost for replace operation: ");
            int replaceCost = scanner.nextInt();
            
            return new OperationCosts(insertCost, deleteCost, replaceCost);
        }
    }
    
    private static int editDistance(String s1, String s2, OperationCosts costs) {
        Scanner scanner = new Scanner(System.in);
        
        System.out.println("Do you want detailed operations? (yes=1 / no=0)");
        System.out.print("Enter your choice (1/0): ");
        int flag = scanner.nextInt();
        
        if (flag == 1) {
            LAResult result = levenshteinCostAndOperation(s1, s2, 0, 0, costs);
            System.out.println("\nOperations to convert '" + s1 + "' to '" + s2 + "':");
            for (String op : result.operations) {
                System.out.println(" - " + op);
            }
            System.out.println();
            return result.cost;
        }
        
        return levenshteinCostOnly(s1, s2, 0, 0, costs);
    }
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        System.out.println("This is a Java program to calculate the Levenshtein distance with custom operation costs.\n\n");
        
        System.out.print("Enter the number of string pairs you want to process: ");
        int numInputs = scanner.nextInt();
        System.out.println();
        
        for (int i = 1; i <= numInputs; i++) {
            System.out.println("\n\n\n ***** ***** ***** ***** Pair " + i + " ***** ***** ***** ***** \n");
            
            // Get operation costs configuration
            OperationCosts costs = getCosts();
            
            System.out.print("Enter the first string (The original string): ");
            String str1 = scanner.next();
            System.out.print("Enter the second string (String to convert to): ");
            String str2 = scanner.next();
            
            System.out.println("\nResults for { Pair " + i + " }");
            int res = editDistance(str1, str2, costs);
            System.out.println("Total Edit Distance for Pair " + i + ": " + res);
        }
        
        scanner.close();
    }
}